### PR TITLE
Add layer loading indicator

### PR DESCRIPTION
**Pull request by reed1**
For issue #463 

In short, this adds a loading icon on top right side indicating map tile images are loading.
Demo URL: http://prism-r2.surge.sh/?hazardLayerId=hydra_s1&date=2022-04-23

Try panning / zooming, the loading indicator will show.

Apologize for the quirky code, but this is the approach that has the minimum performance impact I've found so far. I'll try to breakdown the issues:

1.  Originally I wanted to put `isTileLoading` to `mapStateSlice`, but as I tried, the dispatch for
```
setTileLoading: (state, { payload }: PayloadAction<boolean>) => ({
  ...state,
  isTileLoading: payload
}),
```
costs quite significant in profiler, it causes hiccup on UI. So a new state slice is created just for that map tile loading

2. `map.on('dataloading' | 'idle'...` is used manually instead of using MapboxMap props. This is because `react-mapbox-gl` is not exposing `onIdle` events. Maybe it's worth to fork the `react-mapbox-gl` library, replacing all occurrence of 'mapbox-gl' to 'maplibre-gl' and fix the discrepancies on Interfaces. Also with this, newer maplibre version (2.x) can be supported. I can work on this too if needed.

3. Manual state management on `let tileLoading = false`. The `dataloading` events will be emitted quite often from the map and apparently setting loading state from `true` value to `true` will still have noticeable impact, the scroll isn't as smooth. So the workaround is to monitor the loading manually with local variable, then only dispatch change to state when the local variable changes.

### CODE REVIEW

**Comment by wadhwamatic** [comment]
@reed1  - why aren't we using the existing loading spinner in the center of the page?

@ericboucher  - your thoughts?

**Comment by wadhwamatic** [comment]
> 2\. `map.on('dataloading' | 'idle'...` is used manually instead of using MapboxMap props. This is because `react-mapbox-gl` is not exposing `onIdle` events. Maybe it's worth to fork the `react-mapbox-gl` library, replacing all occurrence of 'mapbox-gl' to 'maplibre-gl' and fix the discrepancies on Interfaces. Also with this, newer maplibre version (2.x) can be supported. I can work on this too if needed.

@DanielJDufour - could you share your input on this?

**Comment by wadhwamatic** [comment]
See issue https://github.com/WFP-VAM/prism-frontend/issues/463

**Comment by reed1** [comment]
> @reed1 - why aren't we using the existing loading spinner in the center of the page?

@wadhwamatic I think it's for loading larger datasets, for loading tile images, the user should still be able to interact with map (panning & zooming), also to minimize distraction as the loading would trigger quite often

**Comment by wadhwamatic** [comment]
@reed1 - I see your point about it being distracting. Another option to explore is to use the legend area and make it specific to the layer which is loading. I.e., it could be an overlay for that layer's section in the legend.

**Comment by ericboucher** [comment]
@reed1 please update the surge links and add a quick video/gif demo when you get a chance.

I think the approach you took works well for now. We can always improve in the future but I think it is efficient and gets the job down, nice work. Ideally, we would want the info to be at the layer level but I understand that it can be tricky and consume a lot of resources, creating lag...

cc @wadhwamatic

File: src/components/MapView/TileLoadingIcon/index.tsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+function TileLoadingIcon() {
+  const loading = useSelector(loadingSelector);
+  if (loading) {
+    return (
+      <div style={{ paddingTop: '0.3rem', paddingRight: '0.3rem' }}>
+        {loading && <CircularProgress size="2rem" />}
+      </div>
+    );
+  }
+  return <></>;
```
Comment (ericboucher) [inline src/components/MapView/TileLoadingIcon/index.tsx:15]: Can you return null instead of an empty element?

**Comment by reed1** [comment]
@ericboucher @wadhwamatic I updated the approach to track loading state per layer. The loading indicator will now show on the legends

image

you can try it on : http://prism-464.surge.sh/?hazardLayerIds=hydra_s1&date=2022-06-29
thanks to automated deploy url feature..

**Review by wadhwamatic — COMMENTED** [review]
Nice work @reed1. I tested in a few scenarios and this worked as expected. Agreed with Eric that we may improve on this in the future (I'm thinking in design terms) but this works well for now. 

Over to @ericboucher for final approval before merging

File: src/components/MapView/Legends/LoadingBar.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  const loading = loadingLayerIds.includes(layerId);
+  return (
+    <div className={classes.container}>
+      <div className={classes.lineBase} />
+      {loading && <div className={classes.line} />}
+    </div>
+  );
+}
+
+const styles = () =>
```
Comment (ericboucher) [inline src/components/MapView/Legends/LoadingBar.tsx:18]: Can we simplify this style a bit now that it's stable? No big deal if not but I feel like some of this css is redundant / unused

**Review by ericboucher — APPROVED** [review]
Great work!
Just a minor comment on simplifying the CSS

File: src/components/MapView/index.tsx [hunk 3]
```diff
@@ -427,6 +427,14 @@ function MapView({ classes }: MapViewProps) {
           dispatch(setLoadingLayerIds([...layerIds]));
         }
       }
+    };
+    map.on('dataloading', listener);
```
Comment (ericboucher) [inline src/components/MapView/index.tsx:431]: Can you explain this new changes?

Comment (wadhwamatic) [reply src/components/MapView/index.tsx:431]: @reed1 - please add some text explaining that last change, then we we can merge. thanks!

**Comment by reed1** [comment]
@ericboucher @wadhwamatic

Deploy URL : http://prism-464.surge.sh/?hazardLayerIds=hydra_s1&date=2022-07-07
Here are the things that's changed.

1. I changed it back to "sourcedataloading" and "sourcedata". I thought some events was not fired, so I used "dataloading" and "data" events with sourceId filtering. But when I checked on maplibre-gl directly, it does fire "sourcedata" and "data" events at the same time. 

2. Minor issue: if we zoom in and immediately zoom out to previous zoom level, the sequence of fired events are:

move - zoom - render - dataloading (isSourceLoaded false) - sourcedataloading (isSourceLoaded false) - moveend - zoomend - render - idle

however there are no events which indicate the layer is now loaded (isSourceLoaded true). So the loading bar will continue forever. The newer version of maplibre-gl has separate sourcedataabort events fired, but also "sourcedata" events will be fired with (isSourceLoaded true) if requests are aborted. But upgrading to newer maplibre-gl is not simple process. I added idle event handler on the bottom of tracking code to workaround this issue. 

3. I removed the `<Grid container direction="column"...` on top Legends component, it seems to have issue with bar width, otherwise I need to hardcode the bar width to 180px, if the container grid is not there, the bar will fit on Legend component correctly

**Comment by ericboucher** [comment]
@reed1 sounds good, can you:
- add more comments in the code
- open an issue to upgrade maplibre and explain what can be changed in your code once it's done

Thanks!

**Comment by reed1** [comment]
@ericboucher sorry which part of it has to be commented ? I think I put several of them already.

As for maplibre, I think the v1.15.2 of maplibre might be the latest that is a drop-in replacement for mapbox-gl. The `sourcedataabort` was introduced in v2.1.1 https://github.com/maplibre/maplibre-gl-js/commit/1e8f75cc66a948eab89274ac139417cd5fa73001 in this PR which has many breaking changes with v1. I don't think it's still a drop-in replacement. I tried to replace build files `(dist/maplibre-gl.js)`, on v1 that is greater than 1.15.2, prism still compiles but has runtime error, on v2 it doesn't compile anymore. If we go with version 2, we might have to refactor the react-mapbox-gl library to sinchronize the interfaces.

**Comment by ericboucher** [comment]
@reed1 ok this PR is go to go, can you open an issue for maplibe? Maybe @DanielJDufour will be able to help if it becomes a blocker

**Comment by reed1** [comment]
@ericboucher I changed a bit on Loading bar, now it will use simple Divider on non-WMS layer. The loading for point data and admin level data is handled differently (using full screen circle progress). Or should we move all loading indicator to legends?

image

I my opinion, there's nothing wrong with the state of maplibre now, the issue is that we want to have a feature that's available on maplibre v2 (newer version) to be backported to v1 that is still a drop-in replacement for mapbox-gl so that it's still compatible with react-mapbox-gl library. Another possibility might be changing the react binding library from react-mapbox-gl (seems not maintained) to react-map-gl (still maintained and supports maplibre gl too: https://visgl.github.io/react-map-gl/docs/get-started/get-started#using-with-a-mapbox-gl-fork ). 

My initial finding for react-map-gl is that they have initialViewState which is less hackish than using `map` reference and call `map.jumpTo` directly to avoid map re-centering on layers change. Other properties seems similar, but they handle GeoJSONLayer a bit different with GeoJSONSource. I don't know if anyone  have looked into that, but seems a viable alternative

**Comment by ericboucher** [comment]
@reed1 thanks for the context, can you actually open a task on GitHub and copy-paste it in there? https://github.com/WFP-VAM/prism-frontend/issues/new Thanks

**Comment by ericboucher** [comment]
@reed1 regarding the design I'd prefer that when it's not loading the bars look the same.
For bar vs full-screen loading, I will let Amit decide @wadhwamatic  but I think am leaning towards using the bar you just developped for most cases.

**Comment by wadhwamatic** [comment]
> @reed1 regarding the design I'd prefer that when it's not loading the bars look the same. For bar vs full-screen loading, I will let Amit decide @wadhwamatic but I think am leaning towards using the bar you just developped for most cases.

Agreed, let's use uniform loading bars

**Comment by wadhwamatic** [comment]
@reed1 & @ericboucher - what's the status of this PR?

**Comment by ericboucher** [comment]
@wadhwamatic waiting for small design changes + using the same bar for other layer loading events

**Comment by reed1** [comment]
@ericboucher @wadhwamatic Now loading indicator is unified in legends:

Deploy URL: http://prism-464.surge.sh/

I'm still using mapStateSlice to track layerData loading, it doesn't incur when people are dragging the map so it's ok to put it there. I'm using the same loadingLayerIds variable naming for both mapStateSlice and mapTileLoadingStateSlice, but refer them as vector and tile on LoadingBar component, hopefully it's not too confusing. Let me know if I need to change anything.

**Comment by ericboucher** [comment]
@reed1 can you resolve the remaining conflict with the master branch?

**Comment by reed1** [comment]
@ericboucher sorry I didn't aware, resolved now

**Pull request merged**